Add a tool to simulate, visualize and compare merge selectors

serxa opened the pull request:
It helps to design and compare merge selector algorithms. Along with other tools, simulations, and visualizations, this PR contains the model that optimally solves the problem of minimizing the average number of parts for a given write-amplification factor value w/o inserts. The model is optimized using the subgradient algorithm to obtain bases (number of parts to merge) for the "Fixed-base" merge selector. It takes into account a limited number of workers in the pool for parallel merging.

<img width="904" height="697" alt="Screenshot 2025-08-06 at 19 59 54" src="https://github.com/user-attachments/assets/fe7ff642-b51e-47eb-8c9c-0c1e740245dc" />

<img width="1358" height="395" alt="Screenshot 2025-08-06 at 19 58 41" src="https://github.com/user-attachments/assets/b1014563-aca1-4cca-a5ac-9e49dd5f6688" />


<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_builds--> Allow: All Builds
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [ ] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
---
- [ ] <!---ci_include_fuzzer--> Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ] <!---ci_exclude_ast--> Exclude: AST fuzzers
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---woolen_wolfdog--> Woolen Wolfdog
- [ ] <!---upload_all--> Upload binaries for special builds
- [ ] <!---no_merge_commit--> Disable merge-commit
- [ ] <!---no_ci_cache--> Disable CI cache

serxa committed:
Add tool to simulate, visualize and compare merge selectors on predefined scenarios

robot-ch-test-poll1 commented:
<!-- automatic status comment for PR #71496 from ClickHouse/ClickHouse:merge-selector-lab -->
*This is an automated comment for commit aceccabfc6f8a8c03cc845d8a4c6f249a990f1e1 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/71496/aceccabfc6f8a8c03cc845d8a4c6f249a990f1e1/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Builds</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/54a1b1c5733e3fdd029aa2499ea7e99430241358/builds/report.html">✅ success</a></td></tr>
<tr><td>ClickBench</td><td>Runs [ClickBench](https://github.com/ClickHouse/ClickBench/) with instant-attach table</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/54a1b1c5733e3fdd029aa2499ea7e99430241358/clickbench__aarch64_.html">✅ success</a></td></tr>
<tr><td>Compatibility check</td><td>Checks that <b>clickhouse</b> binary runs on distributions with old libc versions. If it fails, ask a maintainer for help</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/54a1b1c5733e3fdd029aa2499ea7e99430241358/compatibility_check__aarch64_.html">✅ success</a></td></tr>
<tr><td>Docker keeper image</td><td>The check to build and optionally push the mentioned image to docker hub</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/54a1b1c5733e3fdd029aa2499ea7e99430241358/docker_keeper_image.html">✅ success</a></td></tr>
<tr><td>Docker server image</td><td>The check to build and optionally push the mentioned image to docker hub</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/54a1b1c5733e3fdd029aa2499ea7e99430241358/docker_server_image.html">✅ success</a></td></tr>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/bdd18b35b29f6078b620c65a4f4d59853cfe5371/docs_check.html">✅ success</a></td></tr>
<tr><td>Fast test</td><td>Normally this is the first check that is ran for a PR. It builds ClickHouse and runs most of <a href="https://clickhouse.com/docs/en/development/tests#functional-tests">stateless functional tests</a>, omitting some. If it fails, further checks are not started until it is fixed. Look at the report to see which tests fail, then reproduce the failure locally as described <a href="https://clickhouse.com/docs/en/development/tests#functional-test-locally">here</a></td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/54a1b1c5733e3fdd029aa2499ea7e99430241358/fast_test.html">✅ success</a></td></tr>
<tr><td>Install packages</td><td>Checks that the built packages are installable in a clear environment</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/54a1b1c5733e3fdd029aa2499ea7e99430241358/install_packages__aarch64_.html">✅ success</a></td></tr>
<tr><td>Integration tests</td><td>The integration tests report. In parenthesis the package type is given, and in square brackets are the optional part/total tests</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/54a1b1c5733e3fdd029aa2499ea7e99430241358/integration_tests__aarch64__[1_6].html">✅ success</a></td></tr>
<tr><td>Performance Comparison</td><td>Measure changes in query performance. The performance test report is described in detail <a href="https://github.com/ClickHouse/ClickHouse/tree/master/docker/test/performance-comparison#how-to-read-the-report">here</a>. In square brackets are the optional part/total tests</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/54a1b1c5733e3fdd029aa2499ea7e99430241358/performance_comparison__release__[1_4]/report.html">✅ success</a></td></tr>
<tr><td>Stateful tests</td><td>Runs stateful functional tests for ClickHouse binaries built in various configurations -- release, debug, with sanitizers, etc</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/54a1b1c5733e3fdd029aa2499ea7e99430241358/stateful_tests__aarch64_.html">✅ success</a></td></tr>
<tr><td>Stateless tests</td><td>Runs stateless functional tests for ClickHouse binaries built in various configurations -- release, debug, with sanitizers, etc</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/54a1b1c5733e3fdd029aa2499ea7e99430241358/stateless_tests__aarch64_.html">✅ success</a></td></tr>
<tr><td>Stress test</td><td>Runs stateless functional tests concurrently from several clients to detect concurrency-related errors</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/54a1b1c5733e3fdd029aa2499ea7e99430241358/stress_test__tsan_.html">✅ success</a></td></tr>
<tr><td>Unit tests</td><td>Runs the unit tests for different release types</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/54a1b1c5733e3fdd029aa2499ea7e99430241358/unit_tests__asan_.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

serxa committed:
cleanup Chart.js

serxa committed:
improve axis rescaling, add names

serxa committed:
add tooltips that describe charts

serxa committed:
better tooltips

serxa committed:
add visualization for duration of merges

serxa committed:
more work on visualizations

serxa committed:
better

serxa committed:
remove excessive logging

serxa committed:
extract infoButton module

serxa committed:
refactoring

serxa committed:
refactoring: rename sim into mt

serxa committed:
better

serxa committed:
dont use d3 in MergeTree

serxa committed:
add discrete event simulator

serxa committed:
add MergeTreeMerger

serxa committed:
add MergeTreeInserter

serxa committed:
add part visualization tool

serxa committed:
add zoom support

serxa committed:
add selectors for x and y axis, fix bugs

serxa committed:
add color selector

serxa committed:
dynamic axis ticks format

serxa committed:
cleanup

serxa committed:
rework for parallel merges using event simulator

serxa committed:
add visualization during simulation

serxa committed:
add inserter of parts from clikchouse

serxa committed:
vizualization fixes

serxa committed:
refactor visualizations

serxa committed:
support async generator, fix bugs

serxa committed:
fix

serxa committed:
add realtime timers to EventSimulator

serxa committed:
dynamic updates of visualizations to avoid OOM

serxa committed:
improve visuals

serxa committed:
support merging with inserts + add merge selectors

serxa committed:
show waiting on time diagram

serxa committed:
add time rewinder

serxa committed:
add rewinder buttons for steps

serxa committed:
refactor

serxa committed:
fix intersecting merges

serxa committed:
cleanup

serxa committed:
add small bundler

serxa committed:
more fixes

serxa committed:
bugfix

serxa committed:
add setTime() to rewind simulation

serxa committed:
add gradient descent solver

serxa committed:
add table with optimal results

serxa committed:
better

serxa committed:
more work

serxa committed:
better layer merges on right boundary

serxa committed:
add "Average final part size" input parameter

serxa committed:
refactoring

serxa committed:
working on faster solver

serxa committed:
fast and more accurate solver

serxa committed:
better training page

serxa committed:
solver serialization, solution inference

serxa committed:
add solver test page

serxa committed:
bilinear interpolation

serxa committed:
better chart domains

serxa committed:
faster training + step size + x4 solver

serxa committed:
fix BLI solver

serxa committed:
rename file

serxa committed:
add y-axis options on solution chart

serxa committed:
working on merging plan

serxa committed:
keep working on MergingPlan

serxa committed:
MergingPlan is ready

serxa committed:
system.parts visulalizer is not in #79437

serxa committed:
WIP: unfinished work on complex merge selector

serxa committed:
Revert "WIP: unfinished work on complex merge selector"

This reverts commit 27b1d1ef2a48e2e1162fd6ad87e6c9109f5d3089.

serxa committed:
Revert "MergingPlan is ready"

This reverts commit 88d60b998c4caf848572495e3d76261aa7cc9166.

serxa committed:
Revert "keep working on MergingPlan"

This reverts commit a160df7d5601a3710c94e91c2059dac98bcc73dc.

serxa committed:
Revert "working on merging plan"

This reverts commit 22a8181ac9206cdbb85d8736f27ebdb6973e2eec.

serxa committed:
cleanup and add navigation main page

serxa committed:
buttons for tools

serxa committed:
cleanup, renames

serxa committed:
split scenarios into groups

serxa committed:
adjust scenario descriptions

serxa committed:
cleanup main page

serxa commented:
I've cleaned up unused or uninteresting parts of the research, added the main page for navigation, and added descriptions for every simulation or tool. The optimal model in the presence of inserts or unequal initial parts is still missing and will be the subject of further research.

serxa committed:
fix style
